### PR TITLE
Performance optimizations for StateProofs

### PR DESCRIPTION
**Pull request by Aharonee**
- Update tryBroadcast to run sequentially and break upon failure
- Optimize StateProof Builder to cache the built StateProof and rebuild only when a signature was added

### CODE REVIEW

File: crypto/stateproof/builder.go [hunk 1]
```diff
@@ -46,6 +46,8 @@ type Builder struct {
 	lnProvenWeight uint64
 	provenWeight   uint64
 	strengthTarget uint64
+	stateProof     *StateProof
```
Comment (algoidan) [inline crypto/stateproof/builder.go:49]: 1 should that be `cachedStateProof`?
2 why don't we give up skipBuild and make it a nil ptr when `Add` is invoked?

File: stateproof/builder.go [hunk 2]
```diff
@@ -402,7 +403,14 @@ func (spw *Worker) tryBroadcast() {
 	spw.mu.Lock()
 	defer spw.mu.Unlock()
 
-	for rnd, b := range spw.builders {
+	rounds := make([]basics.Round, 0, len(spw.builders))
+	for rnd := range spw.builders {
+		rounds = append(rounds, rnd)
+	}
+	sort.Slice(rounds, func(i, j int) bool { return rounds[i] < rounds[j] })
```
Comment (algoidan) [inline stateproof/builder.go:410]: why don't we export that into a function? and rename `rounds` to have an hint that it is ordered slice

File: data/transactions/stateproof.go [hunk 3]
```diff
@@ -42,6 +45,31 @@ func (sp StateProofTxnFields) Empty() bool {
 		sp.Message.MsgIsZero()
 }
 
+// String returns a string representation of the StateProofTxnFields.
+func (sp StateProofTxnFields) String() string {
+	return fmt.Sprintf(
```
Comment (algoidan) [inline data/transactions/stateproof.go:50]: I really like the content of the structure - i.e what we choose to print.
few points
- should make sure a field is empty before printing
- should be in one line
- I think we should give up the "keys" just to be aligned with how transaction is being printed.

File: crypto/stateproof/builder_test.go [hunk 4]
```diff
@@ -605,6 +608,28 @@ func TestBuilderWithZeroProvenWeight(t *testing.T) {
 
 }
 
+func TestBuilder_Build(t *testing.T) {
```
Comment (algoidan) [inline crypto/stateproof/builder_test.go:611]: TestBuilder_Build -> invalidateCache

File: stateproof/builder.go (hunk 2 above)
Comment (Aharonee) [reply stateproof/builder.go:410]: It's a one-line function call..

File: stateproof/verify/stateproof.go [hunk 5]
```diff
@@ -136,50 +137,54 @@ func GetProvenWeight(votersHdr bookkeeping.BlockHeader, latestRoundInProofHdr bo
...
-
-	if proto.StateProofInterval == 0 {
-		return fmt.Errorf("rounds = %d: %w", proto.StateProofInterval, errStateProofNotEnabled)
-	}
-
-	if latestRoundInIntervalHdr.Round%basics.Round(proto.StateProofInterval) != 0 {
-		return fmt.Errorf("state proof at %d for non-multiple of %d: %w", latestRoundInIntervalHdr.Round, proto.StateProofInterval, errNotAtRightMultiple)
+// ValidateStateProof validates that the cryptographic proof of the state proof transaction is correct with respect to the headers.
+func ValidateStateProof(txgroup []transactions.SignedTxn, fetcher BlockHeaderFetcher) error {
+	for i := 0; i < len(txgroup); i++ {
```
Comment (algonautshant) [inline stateproof/verify/stateproof.go:142]: How can the stateproof transaction be part of a group with more than 1 transaction?

File: stateproof/verify/stateproof.go [hunk 6]
```diff
@@ -19,24 +19,25 @@ package verify
 import (
 	"errors"
 	"fmt"
-
```
Comment (algonautshant) [inline stateproof/verify/stateproof.go:22]: We should be able to come up with a better file name for this. 
We already have many files named `stateproof.go`, but `stateproof/verify/stateproof.go` is super special.

File: stateproof/verify/stateproof.go [hunk 7]
```diff
@@ -136,50 +137,54 @@ func GetProvenWeight(votersHdr bookkeeping.BlockHeader, latestRoundInProofHdr bo
...
 	}
 
-	err = verifier.Verify(uint64(latestRoundInIntervalHdr.Round), msg.Hash(), stateProof)
+	err = verifier.Verify(uint64(lastRoundInIntervalHdr.Round), stateProofTxn.Message.Hash(), &stateProofTxn.StateProof)
 	if err != nil {
-		return fmt.Errorf("%v: %w", err, errStateProofCrypto)
+		return fmt.Errorf("%v: %w", err, ErrStateProofCrypto)
 	}
 	return nil
+
```
Comment (algonautshant) [inline stateproof/verify/stateproof.go:189]: extra line not needed.

File: stateproof/verify/stateproof.go [hunk 8]
```diff
@@ -19,24 +19,25 @@ package verify
...
 var (
-	errStateProofCrypto                 = errors.New("state proof crypto error")
-	errStateProofParamCreation          = errors.New("state proof param creation error")
-	errStateProofNotEnabled             = errors.New("state proofs are not enabled")
-	errNotAtRightMultiple               = errors.New("state proof is not in a valid round multiple")
-	errInvalidVotersRound               = errors.New("invalid voters round")
-	errExpectedDifferentStateProofRound = errors.New("expected different state proof round")
-	errInsufficientWeight               = errors.New("insufficient state proof weight")
+	ErrStateProofCrypto        = errors.New("state proof crypto error")
+	ErrStateProofParamCreation = errors.New("state proof param creation error")
```
Comment (algonautshant) [inline stateproof/verify/stateproof.go:39]: These do not need to be exported, since they are only used in this package.

**Pull request closed without merging**